### PR TITLE
Part three: album

### PR DESCRIPTION
**Pull request by ltxitai**
Show first 200 Editors choice images, with promotional images every 10
images. Images are cached and can be opened in full screen. Promotional
cells link to Lightricks' website.

### CODE REVIEW

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    [self notifyPendingForURL:requestURL image:image error:error];
+  }];
+
+}
+
+- (void)notifyPendingForURL:(NSURL *)requestURL image:(UIImage * _Nullable)image
+                      error:(NSError * _Nullable)error {
+  @synchronized (self) {
+    NSArray<BGRRemoteImageFetchCompletionHandler> *pending =
+        self.pendingCompletionHandlers[requestURL];
```
Comment (gershonho) [inline Baygur/BGRCachingImageFetchStrategy.mm]: `auto pendingCompletionHandlersForURL = ...`

File: Baygur/BGRImageCell.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2019 Lightricks. All rights reserved.
+// Created by Itai Shaked.
+
+#import "BGRImageDisplayer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class BGRImageMetadata;
```
Comment (gershonho) [inline Baygur/BGRImageCell.h]: Doc

File: Baygur/BGRImageCell.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+/// Cell object for displaying images stored remotely.
+@interface BGRImageCell : UICollectionViewCell <BGRImageDisplayer>
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_UNAVAILABLE;
```
Comment (gershonho) [inline Baygur/BGRImageCell.h]: Doc, empty line

File: Baygur/BGRImageDataSource.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+@property (readonly, nonatomic) NSArray<BGRImageMetadata *> *images;
+
+/// Identifier for the reusable cell type used to display images.
+@property (readonly, nonatomic) NSString *imageCellIdentifier;
+
+/// Identifier for the reusable cell type used to display promotional images.
+@property (readonly, nonatomic) NSString *promotionalCellIdentifier;
+
+/// How many real images should be displayed between promotional images. Must be > 0.
+@property (readwrite, nonatomic) NSUInteger imagesBetweenPromotionalImage;
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: `intervalBetweenPromotionalImages`

File: Baygur/BGRImageDataSource.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+-(instancetype)initWithImages:(NSArray<BGRImageMetadata *> *)images
+          imageCellIdentifier:(NSString *)imageCellIdentifier
+    promotionalCellIdentifier:(NSString *)promotionalCellIdentifier
+                fetchStrategy:(id<BGRRemoteImageFetchStrategy>)fetchStrategy
+    NS_DESIGNATED_INITIALIZER;
+
+/// Array of image metadata objects for images to be displayed.
+@property (readonly, nonatomic) NSArray<BGRImageMetadata *> *images;
+
+/// Identifier for the reusable cell type used to display images.
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: ... to display Editor Choice images

File: Baygur/BGRImageDataSource.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+@property (readonly, nonatomic) NSString *imageCellIdentifier;
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: `editorChoiceCellIdentifier`

File: Baygur/BGRImageDataSource.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+/// Identifier for the reusable cell type used to display promotional images.
+@property (readonly, nonatomic) NSString *promotionalCellIdentifier;
+
+/// How many real images should be displayed between promotional images. Must be > 0.
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.h]: Number of Editor Choice images to be displayed ...

File: Baygur/BGRImageDataSource.mm [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+- (NSInteger)collectionView:(__unused UICollectionView *)collectionView
+     numberOfItemsInSection:(NSInteger)section {
+  LTParameterAssert(section == 0, @"Multiple sections not supported.");
+
+  return self.images.count + (self.images.count / self.imagesBetweenPromotionalImage);
+}
+
+- (BOOL)isIndexOfImage:(NSInteger)index {
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: `isIndexOfEditorChoiceImage`

File: Baygur/BGRImageDataSource.mm [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  return ((index + 1) % (self.imagesBetweenPromotionalImage + 1)) != 0;
+}
+
+- (NSInteger)imageIndexForIndex:(NSInteger)index {
```
Comment (gershonho) [inline Baygur/BGRImageDataSource.mm]: `editorChoiceIndexForGeneralIndex`

File: Baygur/BGRPixabayImageMetadataFetcher.mm [hunk 10]
```diff
@@ -31,13 +31,13 @@ - (instancetype)initWithAPIKey:(NSString *)key baseURLString:(NSString *)urlStri
     _apiKey = [key copy];
     _urlString = [urlString copy];
     _session = [NSURLSession sessionWithConfiguration:
```
Comment (gershonho) [inline Baygur/BGRPixabayImageMetadataFetcher.mm]: Try `sharedSession` - it should be good enough.

File: Baygur/BGRPromoCell.h [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+// Created by Itai Shaked.
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Cell object for displaying a promotional image from the resource bundle. Tapping the cell opens
+/// Lighticks' website. 
+@interface BGRPromoCell : UICollectionViewCell
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_UNAVAILABLE;
```
Comment (gershonho) [inline Baygur/BGRPromoCell.h]: Doc, empty line.

File: Baygur/BGRRemoteImageView.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol BGRRemoteImageFetchStrategy;
+
+/// View for loading and displaying remotely stored images. While loading the view shows an
+/// activity indicator.
+@interface BGRRemoteImageView : UIView
+
+- (instancetype)initWithFrame:(CGRect)frame NS_UNAVAILABLE;
```
Comment (gershonho) [inline Baygur/BGRRemoteImageView.h]: doc

File: Baygur/BGRRemoteImageView.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_UNAVAILABLE;
```
Comment (gershonho) [inline Baygur/BGRRemoteImageView.h]: Doc, empty line

File: Baygur/BGRURLSessionImageFetchStrategy.mm [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+- (void)createURLSession {
+  _session = [NSURLSession
+      sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration];
+}
+
+- (void)fetch:(NSURL *)url
+    completionHandler:(BGRRemoteImageFetchCompletionHandler)completionHandler {
+  @synchronized (self) {
+    [[self.session downloadTaskWithURL:url
+                    completionHandler:^(NSURL * _Nullable location,
```
Comment (gershonho) [inline Baygur/BGRURLSessionImageFetchStrategy.mm]: indent

File: BaygurTests/BGRCachingImageFetchStrategy.mm [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  });
+
+  strategy = [[BGRCachingImageFetchStrategy alloc] initWithFetchStrategy:mockStrategy];
+});
+
+it(@"Should return the image", ^{
+  auto goodURL = [NSURL URLWithString:@"Good"];
+  waitUntil(^(DoneCallback done) {
+    [strategy fetch:goodURL
+  completionHandler:^(NSURL * _Nonnull requestURL, UIImage * _Nullable image,
```
Comment (gershonho) [inline BaygurTests/BGRCachingImageFetchStrategy.mm]: indent - throughout this file

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+- (void)fetch:(NSURL *)url
+    completionHandler:(BGRRemoteImageFetchCompletionHandler)completionHandler {
+  @synchronized (self) {
+    auto cachedImage = [self.cache objectForKey:url];
+    if (cachedImage) {
+      completionHandler(url, cachedImage, nil);
+      return;
+    }
+    auto pending = [self.pendingCompletionHandlers objectForKey:url];
```
[the remark this answers is not in the document]
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.mm]: Done.

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    if (!error && image) {
+      // Only add succesfully fetched images to the cache.
+      if (![self.cache objectForKey:requestURL]) {
+        [self.cache setObject:image forKey:requestURL];
+      }
+    }
+
+    [self notifyPendingForURL:requestURL image:image error:error];
+  }];
+
```
[the remark this answers is not in the document]
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.mm]: Done.

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    // Request not in cache and not pending - submit downstream.
+    [self submitFetch:url completionHandler:completionHandler];
+  }
+}
+
+- (void)submitFetch:(NSURL *)url
+    completionHandler:(BGRRemoteImageFetchCompletionHandler)completionHandler {
+  NSMutableArray<BGRRemoteImageFetchCompletionHandler> *pending =
+      [NSMutableArray arrayWithObject:completionHandler];
+  [self.pendingCompletionHandlers setObject:pending forKey:url];
```
[the remark this answers is not in the document]
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.mm]: Done.

File: Baygur/BGRCachingImageFetchStrategy.mm [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    // Request not in cache and not pending - submit downstream.
+    [self submitFetch:url completionHandler:completionHandler];
+  }
+}
+
+- (void)submitFetch:(NSURL *)url
+    completionHandler:(BGRRemoteImageFetchCompletionHandler)completionHandler {
+  NSMutableArray<BGRRemoteImageFetchCompletionHandler> *pending =
```
[the remark this answers is not in the document]
Comment (ltxitai) [reply Baygur/BGRCachingImageFetchStrategy.mm]: Done.

File: Baygur/BGRImageCell.h (hunk 2 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.h]: Not needed, as discussed.

File: Baygur/BGRImageCell.h (hunk 3 above)
Comment (ltxitai) [reply Baygur/BGRImageCell.h]: Not needed, as discussed.

File: Baygur/BGRImageDataSource.h (hunk 5 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.h (hunk 7 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRPromoCell.h (hunk 11 above)
Comment (ltxitai) [reply Baygur/BGRPromoCell.h]: Not needed, as discussed.

File: Baygur/BGRRemoteImageView.h (hunk 12 above)
Comment (ltxitai) [reply Baygur/BGRRemoteImageView.h]: Not needed, as discussed.

File: Baygur/BGRRemoteImageView.h (hunk 13 above)
Comment (ltxitai) [reply Baygur/BGRRemoteImageView.h]: Not needed, as discussed.

File: Baygur/BGRURLSessionImageFetchStrategy.mm (hunk 14 above)
Comment (ltxitai) [reply Baygur/BGRURLSessionImageFetchStrategy.mm]: Done.

File: Baygur/BGRImageDataSource.h (hunk 4 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.h (hunk 6 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.h]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 8 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRImageDataSource.mm (hunk 9 above)
Comment (ltxitai) [reply Baygur/BGRImageDataSource.mm]: Done.

File: Baygur/BGRPixabayImageMetadataFetcher.mm (hunk 10 above)
Comment (ltxitai) [reply Baygur/BGRPixabayImageMetadataFetcher.mm]: Done.

**Comment by ltxitai** [comment]
Round.

**Comment by gershonho** [comment]
Squash

**Comment by ltxitai** [comment]
Done.

**Pull request merged**